### PR TITLE
Feature/elementwise

### PR DESCRIPTION
**Pull request by campsd**
Implements issues #8 and #7

### CODE REVIEW

File: funfact/lang/interpreter/__init__.py [hunk 1]
```diff
@@ -7,10 +7,13 @@
 from ._index_propagation import IndexPropagator
 from ._initialization import LeafInitializer
 from ._latex import LatexRenderer
+from ._elementwise import ElementEvaluator
```
Comment (yhtang) [inline funfact/lang/interpreter/__init__.py:10]: Should it be called `ElementwiseEvaluator`? Does it feel too long? :laughing:

File: funfact/lang/interpreter/_slicing_propagation.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    def ein(self, lhs: Numeric, rhs: Numeric, precedence: int, reduction: str,
+            pairwise: str, outidx: Optional[P.indices], slices, live_indices,
+            **kwargs):
+        indices_pl = ordered_intersect(live_indices, lhs.live_indices)
+        indices_pr = ordered_intersect(live_indices, rhs.live_indices)
+        indices_none = ordered_setminus(ordered_union(lhs.live_indices,
+                                        rhs.live_indices), live_indices)
+        lhs.slices = (*slices[:len(indices_pl)],
+                      *((slice(None),) * len(indices_none)))
```
Comment (yhtang) [inline funfact/lang/interpreter/_slicing_propagation.py:56]: Using a list here for `slice(None)` could be more concise.

File: funfact/lang/interpreter/_slicing_propagation.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    def neg(self, x: Numeric, slices, **kwargs):
+        x.slices = slices
+
+    def ein(self, lhs: Numeric, rhs: Numeric, precedence: int, reduction: str,
+            pairwise: str, outidx: Optional[P.indices], slices, live_indices,
+            **kwargs):
+        indices_pl = ordered_intersect(live_indices, lhs.live_indices)
+        indices_pr = ordered_intersect(live_indices, rhs.live_indices)
+        indices_none = ordered_setminus(ordered_union(lhs.live_indices,
```
Comment (yhtang) [inline funfact/lang/interpreter/_slicing_propagation.py:53]: I understand `indices_none` refers to the indices that correspond to a range of `slice(None)`. However, the name by itself could be a little confusing at first glance. Could we rename it to something more intuitive?

File: funfact/model/_factorization.py [hunk 4]
```diff
@@ -47,6 +50,17 @@ def forward(self):
...
+    def getelements(self, idx):
+        '''Get elements at index of tensor expression.'''
+        new_idx = []
+        for i in idx:
+            if isinstance(i, int):
+                new_idx.append(slice(i, i+1))
+            else:
+                new_idx.append(i)
+        _index_slicer = SlicingPropagator(tuple(new_idx))
+        return (self.tsrex | _index_slicer) | self._element_evaluator
```
Comment (yhtang) [inline funfact/model/_factorization.py:62]: I believe we can get rid of the parentheses here.

File: funfact/model/_factorization.py [hunk 5]
```diff
@@ -47,6 +50,17 @@ def forward(self):
         '''Evaluate the tensor expression the result.'''
         return self.tsrex | self._evaluator
 
+    def getelements(self, idx):
```
Comment (yhtang) [inline funfact/model/_factorization.py:53]: It's better to name it `_get_elements` since we don't want users to call it directly.

**Review by yhtang — CHANGES_REQUESTED** [review]

**Comment by yhtang** [comment]
Great job! I didn't realize the new evaluator itself needs only 3 lines. lol

**Comment by campsd** [comment]
@yhtang  I have fixed the index ordering bug and added an extra test for Tucker decomposition.

**Review by yhtang — APPROVED** [review]

**Pull request merged**